src/1-theory/10-async.adoc

```diff
@@ -0,0 +1,110 @@
+
+.Letting the API do a bit less work up front, shiting the waiting work to a job queue, and letting the GUI check for updates.
+[plantuml]
+----
+@startuml
+
+title Long Polling and Job Queues
+
+actor User
+boundary "Web GUI" as GUI
+control "API"
+control Worker
+boundary "Payment Gateway" as PG
+
+User -> API: POST /payments
+API -> Worker: Queue Job
+GUI <- API: HTTP 202 Accepted (with a URL to poll)
+User <- GUI: User sees "payment in progress"
+GUI -> GUI: Sets a timer to start calling the API every X seconds
+
+alt payment gateway not ready
+    GUI -> API: GET /payments/abc
+    GUI <- API: HTTP 200 OK { status: pending }
+else payment gateway has approved the payment
+    Worker --> PG: POST /submitPaymentAttempt
+    PG -> PG: May take 30-60s to confirm
+    Worker <- PG: HTTP 200 OK
+    Worker -> API: Update payment status
+    GUI -> API: GET /payments/abc
+    GUI <- API: HTTP 200 OK { status: approved }
+    User <- GUI: User gets confirmation
+end
+
+@enduml
+----
+
+This is the very first step in using APIs asynchronously in your architecture, and job queues are amazingly useful for more than just this especially in the world of microservices. We will get to that, but first, what is wrong with the approach in the diagram?
+
+== Long Polling
+
+The client is having to consistently ask if the thing is ready, which... is annoying. Making HTTP requests you don't need to make is going to drain the battery of mobile devices, add unnecessary load to the API server, and is usually pretty avoidable.
+
+Some people think that long polling is inherintly part of how some APIs (like REST) are meant to work, but there are loads of alternatives to this very primitive paradigm. For starters, Web Hooks (a.k.a Callbacks)
+
+== Web Hooks / Callbacks
+
+These are two pretty common names for the same thing. The idea is, a client can register a URL for an API to fire a payload at when updates happen to something they are interested in. This usually takes the form of a HTTP POST request, and is common in most APIs.
+
+- https://api.slack.com/incoming-webhooks[Slack API > Incoming Webhooks]
+- https://stripe.com/docs/webhooks[Stripe API > Webhooks]
+
+Webhooks are great for integrating server-to-server, where an API wants to let another server (maybe an API, maybe just an application) know that something has happened. This would improve our user experience if the GUI in our example was a server-side backed application, but if it is a JavaScript application sitting in the users browser we cannot just fire stuff at their IP address and hope it works (it won't).
+
+Callbacks do however improve the right hand side of this diagram a bit, if our payment gateway provider is using them.
+
+Thankfully most payment gateways do not work like this. For example, Stripe will respond with a qiuck response (this information looks fairly valid) but we will let you know if the bank approved the transaction for this card.
```

Comment:
```suggestion
Thankfully most payment gateways do not work like this. For example, Stripe will respond with a quick response (this information looks fairly valid) but we will let you know if the bank approved the transaction for this card.
```

Reply:
Thanks @cebe, ignore this until it's in review or you'll have a looooot of typo suggestions. :)

If there is anything factually incorrect, or things that look finished but dont make any sense, maybe call that out.

Reply:
sure, will do. did not spot anything like that.

Reply:
Ready!